File: lib/api-client.ts
// Real Backend API Client for Admin Panel
// Connects to the deployed backend at https://champ-y6eg.onrender.com

const API_BASE_URL = process.env.NEXT_PUBLIC_API_URL || 'https://champ-y6eg.onrender.com/api';

// Admin credentials - Real admin user in database
// Password: Admin@123 (for real backend login)
// Demo password: admin123 (for offline demo mode)
const DEMO_ADMIN = {
  email: 'admin@estato.com',
  password: 'admin123', // Demo password for offline mode
  realPassword: 'Admin@123', // Real password for backend
  token: 'demo-admin-token-12345',
  user: {
    id: '27144234-ceca-4721-92b0-a1b392ad0d55',
    email: 'admin@estato.com',
    name: 'Estato Admin',
    role: 'admin',
    user_type: 'admin',
  }
};

interface ApiResponse<T = any> {
  success: boolean;
  data?: T;
  error?: string;
  message?: string;
  count?: number;
}

class ApiClient {
  private token: string | null = null;

  setToken(token: string) {
    this.token = token;
    if (typeof window !== 'undefined') {
      localStorage.setItem('admin_token', token);
    }
  }

  getToken(): string | null {
    if (this.token) return this.token;
    if (typeof window !== 'undefined') {
      return localStorage.getItem('admin_token');
    }
    return null;
  }

  clearToken() {
    this.token = null;
    if (typeof window !== 'undefined') {
      localStorage.removeItem('admin_token');
    }
  }

  private async request<T>(
    endpoint: string,
    options: RequestInit = {}
  ): Promise<ApiResponse<T>> {
    const token = this.getToken();
    
    const headers: HeadersInit = {
      'Content-Type': 'application/json',
      ...options.headers,
    };

    if (token) {
      (headers as Record<string, string>)['Authorization'] = `Bearer ${token}`;
    }

    try {
      const response = await fetch(`${API_BASE_URL}${endpoint}`, {
        ...options,
        headers,
      });

      const data = await response.json();
      return data;
    } catch (error) {
      console.error('API request failed:', error);
      return {
        success: false,
        error: 'Network error. Please try again.',
      };
    }
  }

  // Auth
  async login(email: string, password: string) {
    // Check for demo credentials (offline mode)
    if (email === DEMO_ADMIN.email && password === DEMO_ADMIN.password) {
      this.setToken(DEMO_ADMIN.token);
      return {
        success: true,
        data: {
          token: DEMO_ADMIN.token,
          accessToken: DEMO_ADMIN.token,
          user: DEMO_ADMIN.user,
        },
        message: 'Demo login successful',
      };
    }

    // Try real backend login
    try {
      const response = await this.request<{ token: string; accessToken: string; user: any }>('/auth/login', {
        method: 'POST',
        body: JSON.stringify({ email, password }),
      });
      
      if (response.success && response.data) {
        const token = response.data.accessToken || response.data.token;
        if (token) {
          this.setToken(token);
        }
        
        // For admin user, also set admin role in response
        if (email === DEMO_ADMIN.email) {
          response.data.user = {
            ...response.data.user,
            role: 'admin',
            user_type: 'admin',
          };
        }
      }
      
      return response;
    } catch (error) {
      console.error('Login API error:', error);
      // If backend is unreachable and using admin email with either password, allow demo login
      if (email === DEMO_ADMIN.email && (password === DEMO_ADMIN.password || password === DEMO_ADMIN.realPassword)) {
        this.setToken(DEMO_ADMIN.token);
        return {
          success: true,
          data: {
            token: DEMO_ADMIN.token,
            user: DEMO_ADMIN.user,
          },
          message: 'Demo login successful (offline mode)',
        };
      }
      return {
        success: false,
        error: 'Unable to connect to server. Please try again.',
      };
    }
  }

  async logout() {
    this.clearToken();
    return { success: true };
  }

  async getMe() {
    // Check for demo token
    const token = this.getToken();
    if (token === DEMO_ADMIN.token) {
      return {
        success: true,
        data: DEMO_ADMIN.user,
      };
    }
    
    return this.request('/auth/me');
  }

  // Dashboard Stats
  async getDashboardStats() {
    return this.request('/admin/stats');
  }

  async getDashboard() {
    return this.request('/admin/dashboard');
  }

  // Analytics
  async getAnalytics(timeRange: string = '30d') {
    return this.request(`/admin/analytics?timeRange=${timeRange}`);
  }

  // Properties - Try admin endpoint first, fall back to public
  async getProperties(status?: string) {
    const query = status ? `?status=${status}` : '';
    
    // Try admin endpoint first (requires auth)
    const adminResponse = await this.request(`/admin/properties${query}`);
    if (adminResponse.success) {
      return adminResponse;
    }
    
    // Fall back to public endpoint with showAll flag
    return this.request(`/properties?showAll=true${status ? `&status=${status}` : ''}`);
  }

  // Get all properties including pending (for admin view)
  async getAllProperties() {
    // Try admin endpoint
    const adminResponse = await this.request('/admin/properties');
    if (adminResponse.success) {
      return adminResponse;
    }
    
    // Fall back - fetch from public with all statuses
    return this.request('/properties?showAll=true');
  }

  async getProperty(id: string) {
    return this.request(`/properties/${id}`);
  }

  async approveProperty(id: string) {
    return this.request(`/admin/properties/${id}/approve`, {
      method: 'PUT',
    });
  }

  async rejectProperty(id: string, reason?: string) {
    return this.request(`/admin/properties/${id}/reject`, {
      method: 'PUT',
      body: JSON.stringify({ reason }),
    });
  }

  async addPropertyComment(id: string, comment: string, status?: string) {
    return this.request(`/admin/properties/${id}/comment`, {
      method: 'PUT',
      body: JSON.stringify({ comment, status }),
    });
  }

  async updatePropertyStatus(id: string, status: string) {
    return this.request(`/admin/properties/${id}/status`, {
      method: 'PUT',
      body: JSON.stringify({ status }),
    });
  }

  async deleteProperty(id: string) {
    return this.request(`/admin/properties/${id}`, {
      method: 'DELETE',
    });
  }

  // Users
  async getUsers() {
    return this.request('/admin/users');
  }

  async verifyUser(id: string) {
    return this.request(`/admin/users/${id}/verify`, {
      method: 'PUT',
    });
  }

  async suspendUser(id: string) {
    return this.request(`/admin/users/${id}/suspend`, {
      method: 'PUT',
    });
  }

  async activateUser(id: string) {
    return this.request(`/admin/users/${id}/activate`, {
      method: 'PUT',
    });
  }

  async updateUserRole(id: string, role: string) {
    return this.request(`/admin/users/${id}/role`, {
      method: 'PUT',
      body: JSON.stringify({ role }),
    });
  }

  async deleteUser(id: string) {
    return this.request(`/admin/users/${id}`, {
      method: 'DELETE',
    });
  }

  // Agents
  async getAgents() {
    return this.request('/admin/agents');
  }

  async approveAgent(id: string) {
    return this.request(`/admin/agents/${id}/approve`, {
      method: 'PUT',
    });
  }

  async rejectAgent(id: string, reason?: string) {
    return this.request(`/admin/agents/${id}/reject`, {
      method: 'PUT',
      body: JSON.stringify({ reason }),
    });
  }

  // Bookings
  async getBookings(status?: string) {
    const query = status ? `?status=${status}` : '';
    return this.request(`/admin/bookings${query}`);
  }

  async updateBookingStatus(id: string, status: string) {
    return this.request(`/admin/bookings/${id}/status`, {
      method: 'PUT',
      body: JSON.stringify({ status }),
    });
  }

  // Reports
  async getReports() {
    return this.request('/admin/reports');
  }

  async resolveReport(id: string, notes?: string) {
    return this.request(`/admin/reports/${id}/resolve`, {
      method: 'PUT',
      body: JSON.stringify({ resolution_notes: notes }),
    });
  }

  async dismissReport(id: string, notes?: string) {
    return this.request(`/admin/reports/${id}/dismiss`, {
      method: 'PUT',
      body: JSON.stringify({ resolution_notes: notes }),
    });
  }

  // Public Properties (no auth required)
  async getPublicProperties(filters?: Record<string, string>) {
    const params = new URLSearchParams(filters);
    return this.request(`/properties?${params.toString()}`);
  }

  async getFeaturedProperties() {
    return this.request('/properties/featured');
  }
}

export const apiClient = new ApiClient();
export default apiClient;

File: lib/services/api_client.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:http_parser/http_parser.dart';
import 'package:shared_preferences/shared_preferences.dart';
import 'config_service.dart';

/// API Client for backend communication
/// 
/// This service handles all API calls to the Node.js backend with Supabase authentication
class ApiClient {
  // Use config service for base URL
  static String get baseUrl => ConfigService.apiBaseUrl;
  
  // Get stored access token
  static Future<String?> _getAccessToken() async {
    final prefs = await SharedPreferences.getInstance();
    final token = prefs.getString('accessToken');
    return token;
  }

  // Save access token
  static Future<void> _saveAccessToken(String token) async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.setString('accessToken', token);
  }

  // Clear access token
  static Future<void> _clearAccessToken() async {
    final prefs = await SharedPreferences.getInstance();
    await prefs.remove('accessToken');
  }

  // Make authenticated request
  static Future<Map<String, dynamic>> _makeRequest(
    String method,
    String endpoint, {
    Map<String, dynamic>? body,
    Map<String, String>? headers,
    bool requireAuth = true,
  }) async {
    try {
      final url = Uri.parse('$baseUrl$endpoint');
      
      final requestHeaders = <String, String>{
        'Content-Type': 'application/json',
        ...?headers,
      };

      if (requireAuth) {
        final token = await _getAccessToken();
        if (token == null) {
          return {
            'success': false,
            'error': 'Not authenticated. Please login.',
          };
        }
        requestHeaders['Authorization'] = 'Bearer $token';
      }

      http.Response response;

      switch (method.toUpperCase()) {
        case 'GET':
          response = await http.get(url, headers: requestHeaders).timeout(
            const Duration(seconds: 30),
          );
          break;
        case 'POST':
          response = await http.post(
            url,
            headers: requestHeaders,
            body: body != null ? json.encode(body) : null,
          ).timeout(
            const Duration(seconds: 30),
          );
          break;
        case 'PUT':
          response = await http.put(
            url,
            headers: requestHeaders,
            body: body != null ? json.encode(body) : null,
          );
          break;
        case 'DELETE':
          response = await http.delete(url, headers: requestHeaders);
          break;
        default:
          throw Exception('Unsupported HTTP method: $method');
      }

      final responseData = json.decode(response.body) as Map<String, dynamic>;

      if (response.statusCode == 401) {
        // Token expired, clear and return error
        await _clearAccessToken();
        return {
          'success': false,
          'error': 'Session expired. Please login again.',
        };
      }

      return responseData;
    } catch (e) {
      String errorMessage = 'Network error';
      
      if (e.toString().contains('TimeoutException')) {
        errorMessage = 'Request timeout. Please check your internet connection.';
      } else if (e.toString().contains('SocketException')) {
        errorMessage = 'Cannot connect to server. Please check your internet connection.';
      } else if (e.toString().contains('HandshakeException')) {
        errorMessage = 'SSL/TLS connection failed. Please try again.';
      } else if (e.toString().contains('FormatException')) {
        errorMessage = 'Invalid response from server.';
      } else {
        errorMessage = 'Network error: ${e.toString()}';
      }
      
      return {
        'success': false,
        'error': errorMessage,
      };
    }
  }

  // Authentication Methods
  static Future<Map<String, dynamic>> register({
    required String email,
    required String password,
    required String name,
    required String phone,
    required String userType,
  }) async {
    final response = await _makeRequest(
      'POST',
      '/auth/register',
      body: {
        'email': email,
        'password': password,
        'name': name,
        'phone': phone,
        'userType': userType.toLowerCase(), // Valid: buyer, seller, agent, landlord, owner, both
      },
      requireAuth: false,
    );

    if (response['success'] == true && response['data'] != null) {
      // Handle both session and direct token response formats
      final data = response['data'];
      String? accessToken;
      
      if (data['session'] != null && data['session']['access_token'] != null) {
        accessToken = data['session']['access_token'];
      } else if (data['accessToken'] != null) {
        accessToken = data['accessToken'];
      }
      
      if (accessToken != null) {
        await _saveAccessToken(accessToken);
      }
    }

    return response;
  }

  static Future<Map<String, dynamic>> login({
    required String email,
    required String password,
  }) async {
    final response = await _makeRequest(
      'POST',
      '/auth/login',
      body: {
        'email': email,
        'password': password,
      },
      requireAuth: false,
    );

    if (response['success'] == true && response['data'] != null) {
      final accessToken = response['data']['accessToken'];
      if (accessToken != null) {
        await _saveAccessToken(accessToken);
      }
    }

    return response;
  }

  static Future<Map<String, dynamic>> logout() async {
    final response = await _makeRequest('POST', '/auth/logout');
    await _clearAccessToken();
    return response;
  }

  static Future<Map<String, dynamic>> getCurrentUser() async {
    return await _makeRequest('GET', '/auth/me');
  }

  static Future<Map<String, dynamic>> refreshToken(String refreshToken) async {
    final response = await _makeRequest(
      'POST',
      '/auth/refresh',
      body: {'refreshToken': refreshToken},
      requireAuth: false,
    );

    if (response['success'] == true && response['data'] != null) {
      final accessToken = response['data']['accessToken'];
      if (accessToken != null) {
        await _saveAccessToken(accessToken);
      }
    }

    return response;
  }

  // OTP Methods
  static Future<Map<String, dynamic>> sendOTP({
    String? phone,
    String? email,
  }) async {
    return await _makeRequest(
      'POST',
      '/otp/send',
      body: {
        if (phone != null) 'phone': phone,
        if (email != null) 'email': email,
      },
      requireAuth: false,
    );
  }

  static Future<Map<String, dynamic>> verifyOTP({
    String? phone,
    String? email,
    required String otp,
  }) async {
    return await _makeRequest(
      'POST',
      '/otp/verify',
      body: {
        if (phone != null) 'phone': phone,
        if (email != null) 'email': email,
        'otp': otp,
      },
      requireAuth: false,
    );
  }

  static Future<Map<String, dynamic>> resendOTP({
    String? phone,
    String? email,
  }) async {
    return await _makeRequest(
      'POST',
      '/otp/resend',
      body: {
        if (phone != null) 'phone': phone,
        if (email != null) 'email': email,
      },
      requireAuth: false,
    );
  }

  // User Methods
  static Future<Map<String, dynamic>> getUserProfile() async {
    return await _makeRequest('GET', '/users/profile');
  }

  static Future<Map<String, dynamic>> updateUserProfile({
    String? name,
    String? phone,
    String? bio,
  }) async {
    return await _makeRequest(
      'PUT',
      '/users/profile',
      body: {
        if (name != null) 'name': name,
        if (phone != null) 'phone': phone,
        if (bio != null) 'bio': bio,
      },
    );
  }

  static Future<Map<String, dynamic>> uploadAvatar(String imagePath) async {
    try {
      final token = await _getAccessToken();
      if (token == null) {
        return {
          'success': false,
          'error': 'Not authenticated. Please login.',
        };
      }

      final url = Uri.parse('$baseUrl/users/avatar');
      final request = http.MultipartRequest('POST', url);
      
      // Add authorization header
      request.headers['Authorization'] = 'Bearer $token';
      
      // Add the file
      final file = await http.MultipartFile.fromPath('avatar', imagePath);
      request.files.add(file);
      
      final streamedResponse = await request.send().timeout(
        const Duration(seconds: 60),
      );
      
      final response = await http.Response.fromStream(streamedResponse);
      final responseData = json.decode(response.body) as Map<String, dynamic>;
      
      if (response.statusCode == 401) {
        await _clearAccessToken();
        return {
          'success': false,
          'error': 'Session expired. Please login again.',
        };
      }
      
      return responseData;
    } catch (e) {
      return {
        'success': false,
        'error': 'Failed to upload avatar. Please try again.',
      };
    }
  }

  static Future<Map<String, dynamic>> changePassword({
    required String currentPassword,
    required String newPassword,
  }) async {
    return await _makeRequest(
      'POST',
      '/users/change-password',
      body: {
        'currentPassword': currentPassword,
        'newPassword': newPassword,
      },
    );
  }

  static Future<Map<String, dynamic>> deleteAccount() async {
    final response = await _makeRequest('DELETE', '/users/profile');
    if (response['success'] == true) {
      await _clearAccessToken();
    }
    return response;
  }

  // Property Methods
  static Future<Map<String, dynamic>> getProperties({
    String? propertyType,
    String? transactionType,
    double? minPrice,
    double? maxPrice,
    String? area,
    String? search,
  }) async {
    final queryParams = <String, String>{};
    if (propertyType != null) queryParams['propertyType'] = propertyType;
    if (transactionType != null) queryParams['transactionType'] = transactionType;
    if (minPrice != null) queryParams['minPrice'] = minPrice.toString();
    if (maxPrice != null) queryParams['maxPrice'] = maxPrice.toString();
    if (area != null) queryParams['area'] = area;
    if (search != null) queryParams['search'] = search;

    final queryString = queryParams.entries
        .map((e) => '${Uri.encodeComponent(e.key)}=${Uri.encodeComponent(e.value)}')
        .join('&');

    return await _makeRequest(
      'GET',
      '/properties${queryString.isNotEmpty ? '?$queryString' : ''}',
      requireAuth: false,
    );
  }

  static Future<Map<String, dynamic>> getUserProperties() async {
    return await _makeRequest('GET', '/users/properties');
  }

  static Future<Map<String, dynamic>> getPropertyById(String id) async {
    return await _makeRequest('GET', '/properties/$id', requireAuth: false);
  }

  static Future<Map<String, dynamic>> createProperty(Map<String, dynamic> propertyData) async {
    try {
      final token = await _getAccessToken();
      if (token == null) {
        return {
          'success': false,
          'error': 'Not authenticated. Please login.',
        };
      }

      final url = Uri.parse('$baseUrl/properties');
      final request = http.MultipartRequest('POST', url);

      // Add authorization header
      request.headers['Authorization'] = 'Bearer $token';

      // Add text fields
      if (propertyData['title'] != null) {
        request.fields['title'] = propertyData['title'];
      }
      if (propertyData['description'] != null) {
        request.fields['description'] = propertyData['description'];
      }
      if (propertyData['price'] != null) {
        request.fields['price'] = propertyData['price'].toString();
      }
      if (propertyData['propertyType'] != null) {
        request.fields['propertyType'] = propertyData['propertyType'];
      }
      if (propertyData['transactionType'] != null) {
        request.fields['transactionType'] = propertyData['transactionType'];
      }
      if (propertyData['location'] != null) {
        request.fields['location'] = propertyData['location'];
      }
      if (propertyData['area'] != null) {
        request.fields['area'] = propertyData['area'];
      }
      if (propertyData['size'] != null) {
        request.fields['size'] = propertyData['size'].toString();
      }
      if (propertyData['bedrooms'] != null) {
        request.fields['bedrooms'] = propertyData['bedrooms'].toString();
      }
      if (propertyData['bathrooms'] != null) {
        request.fields['bathrooms'] = propertyData['bathrooms'].toString();
      }
      if (propertyData['ownerPhone'] != null) {
        request.fields['ownerPhone'] = propertyData['ownerPhone'];
      }
      if (propertyData['isFurnished'] != null) {
        request.fields['isFurnished'] = propertyData['isFurnished'].toString();
      }
      if (propertyData['yearBuilt'] != null) {
        request.fields['yearBuilt'] = propertyData['yearBuilt'].toString();
      }
      if (propertyData['latitude'] != null) {
        request.fields['latitude'] = propertyData['latitude'].toString();
      }
      if (propertyData['longitude'] != null) {
        request.fields['longitude'] = propertyData['longitude'].toString();
      }
      if (propertyData['isFeatured'] != null) {
        request.fields['isFeatured'] = propertyData['isFeatured'].toString();
      }
      if (propertyData['amenities'] != null && propertyData['amenities'] is List) {
        request.fields['amenities'] = json.encode(propertyData['amenities']);
      }

      // Add image files
      if (propertyData['images'] != null && propertyData['images'] is List) {
        final images = propertyData['images'] as List;
        for (int i = 0; i < images.length && i < 10; i++) {
          final imagePath = images[i];
          if (imagePath is String && imagePath.isNotEmpty) {
            // Check if it's a file path (local file) or URL (already uploaded)
            if (!imagePath.startsWith('http')) {
              // It's a local file path, add as multipart file
              try {
                // Determine content type from file extension
                String contentType = 'image/jpeg'; // default
                final lowerPath = imagePath.toLowerCase();
                if (lowerPath.endsWith('.png')) {
                  contentType = 'image/png';
                } else if (lowerPath.endsWith('.gif')) {
                  contentType = 'image/gif';
                } else if (lowerPath.endsWith('.webp')) {
                  contentType = 'image/webp';
                } else if (lowerPath.endsWith('.heic') || lowerPath.endsWith('.heif')) {
                  contentType = 'image/heic';
                }
                
                final file = await http.MultipartFile.fromPath(
                  'images', 
                  imagePath,
                  contentType: MediaType.parse(contentType),
                );
                request.files.add(file);
              } catch (e) {
                // Skip invalid files
                print('Error adding image file: $e');
                continue;
              }
            }
          }
        }
      }

      final streamedResponse = await request.send().timeout(
        const Duration(seconds: 60),
      );

      final response = await http.Response.fromStream(streamedResponse);
      final responseData = json.decode(response.body) as Map<String, dynamic>;

      if (response.statusCode == 401) {
        await _clearAccessToken();
        return {
          'success': false,
          'error': 'Session expired. Please login again.',
        };
      }

      return responseData;
    } catch (e) {
      return {
        'success': false,
        'error': 'Failed to create property. Please try again.',
      };
    }
  }

  static Future<Map<String, dynamic>> updateProperty(String id, Map<String, dynamic> updates) async {
    return await _makeRequest('PUT', '/properties/$id', body: updates);
  }

  static Future<Map<String, dynamic>> deleteProperty(String id) async {
    return await _makeRequest('DELETE', '/properties/$id');
  }

  static Future<Map<String, dynamic>> getSimilarProperties(String propertyId) async {
    return await _makeRequest('GET', '/properties/$propertyId/similar', requireAuth: false);
  }

  // Favorite Methods
  static Future<Map<String, dynamic>> getFavorites() async {
    return await _makeRequest('GET', '/favorites');
  }

  static Future<Map<String, dynamic>> getFavoriteIds() async {
    return await _makeRequest('GET', '/favorites/ids');
  }

  static Future<Map<String, dynamic>> checkFavorite(String propertyId) async {
    return await _makeRequest('GET', '/favorites/check/$propertyId');
  }

  static Future<Map<String, dynamic>> addFavorite(String propertyId) async {
    return await _makeRequest('POST', '/favorites', body: {'propertyId': propertyId});
  }

  static Future<Map<String, dynamic>> toggleFavorite(String propertyId) async {
    return await _makeRequest('POST', '/favorites/toggle', body: {'propertyId': propertyId});
  }

  static Future<Map<String, dynamic>> removeFavorite(String propertyId) async {
    return await _makeRequest('DELETE', '/favorites/$propertyId');
  }

  // Chat Methods
  static Future<Map<String, dynamic>> getChats() async {
    return await _makeRequest('GET', '/chats');
  }

  static Future<Map<String, dynamic>> createChat({
    required String participant2Id,
    required String propertyId,
  }) async {
    return await _makeRequest('POST', '/chats', body: {
      'participant2Id': participant2Id,
      'propertyId': propertyId,
    });
  }

  static Future<Map<String, dynamic>> getChatMessages(String chatId) async {
    return await _makeRequest('GET', '/chats/$chatId/messages');
  }

  static Future<Map<String, dynamic>> sendMessage(String chatId, String content) async {
    return await _makeRequest('POST', '/chats/$chatId/messages', body: {
      'content': content,
    });
  }

  static Future<Map<String, dynamic>> getPropertyContact(String propertyId) async {
    return await _makeRequest('GET', '/chats/contact/$propertyId');
  }

  static Future<Map<String, dynamic>> startChatWithOwner(String propertyId, {String? message}) async {
    return await _makeRequest('POST', '/chats/start/$propertyId', body: {
      if (message != null) 'message': message,
    });
  }

  // Booking Methods
  static Future<Map<String, dynamic>> getBookings({String? status}) async {
    final query = status != null ? '?status=$status' : '';
    return await _makeRequest('GET', '/bookings$query');
  }

  static Future<Map<String, dynamic>> createBooking({
    required String propertyId,
    required String scheduledDate,
    required String scheduledTime,
    String? notes,
  }) async {
    return await _makeRequest('POST', '/bookings', body: {
      'propertyId': propertyId,
      'scheduledDate': scheduledDate,
      'scheduledTime': scheduledTime,
      if (notes != null) 'notes': notes,
    });
  }

  static Future<Map<String, dynamic>> updateBooking(String id, String status) async {
    return await _makeRequest('PUT', '/bookings/$id', body: {'status': status});
  }

  // Admin Methods
  static Future<Map<String, dynamic>> getAdminDashboard() async {
    return await _makeRequest('GET', '/admin/dashboard');
  }

  static Future<Map<String, dynamic>> getAdminUsers() async {
    return await _makeRequest('GET', '/admin/users');
  }

  static Future<Map<String, dynamic>> getAdminProperties({String? status}) async {
    final query = status != null ? '?status=$status' : '';
    return await _makeRequest('GET', '/admin/properties$query');
  }

  static Future<Map<String, dynamic>> approveProperty(String id) async {
    return await _makeRequest('PUT', '/admin/properties/$id/approve');
  }

  static Future<Map<String, dynamic>> rejectProperty(String id) async {
    return await _makeRequest('PUT', '/admin/properties/$id/reject');
  }

  static Future<Map<String, dynamic>> getAdminAgents() async {
    return await _makeRequest('GET', '/admin/agents');
  }

  static Future<Map<String, dynamic>> approveAgent(String id) async {
    return await _makeRequest('PUT', '/admin/agents/$id/approve');
  }

  static Future<Map<String, dynamic>> getAdminReports() async {
    return await _makeRequest('GET', '/admin/reports');
  }

  static Future<Map<String, dynamic>> resolveReport(String id) async {
    return await _makeRequest('PUT', '/admin/reports/$id/resolve');
  }

  // Authentication - Password Reset Methods
  static Future<Map<String, dynamic>> forgotPassword(String email) async {
    return await _makeRequest(
      'POST',
      '/auth/forgot-password',
      body: {'email': email},
      requireAuth: false,
    );
  }

  static Future<Map<String, dynamic>> resetPassword({
    required String password,
    required String token,
  }) async {
    return await _makeRequest(
      'POST',
      '/auth/reset-password',
      body: {
        'password': password,
        'token': token,
      },
      requireAuth: false,
    );
  }

  // Notification Methods
  static Future<Map<String, dynamic>> getNotifications() async {
    return await _makeRequest('GET', '/notifications');
  }

  static Future<Map<String, dynamic>> markNotificationRead(String id) async {
    return await _makeRequest('PUT', '/notifications/$id/read');
  }

  static Future<Map<String, dynamic>> updateNotificationSettings(
    Map<String, dynamic> settings,
  ) async {
    return await _makeRequest(
      'PUT',
      '/notifications/settings',
      body: settings,
    );
  }

  // Settings & Privacy Methods
  static Future<Map<String, dynamic>> updatePrivacySettings(
    Map<String, dynamic> settings,
  ) async {
    return await _makeRequest(
      'PUT',
      '/users/privacy-settings',
      body: settings,
    );
  }

  static Future<Map<String, dynamic>> setup2FA(
    Map<String, String> questionsAndAnswers,
  ) async {
    return await _makeRequest(
      'POST',
      '/auth/setup-2fa',
      body: {'securityQuestions': questionsAndAnswers},
    );
  }

  static Future<Map<String, dynamic>> verify2FA(
    String question,
    String answer,
  ) async {
    return await _makeRequest(
      'POST',
      '/auth/verify-2fa',
      body: {'question': question, 'answer': answer},
    );
  }

  // Login Activity Methods
  static Future<Map<String, dynamic>> getLoginActivities() async {
    return await _makeRequest('GET', '/users/login-activities');
  }

  static Future<Map<String, dynamic>> logoutFromDevice(String activityId) async {
    return await _makeRequest('DELETE', '/users/login-activities/$activityId');
  }

  static Future<Map<String, dynamic>> recordLoginActivity({
    required String deviceName,
    required String deviceType,
    String? location,
    String? ipAddress,
  }) async {
    return await _makeRequest(
      'POST',
      '/users/login-activities',
      body: {
        'deviceName': deviceName,
        'deviceType': deviceType,
        'location': location,
        'ipAddress': ipAddress,
      },
    );
  }

  // Identity Verification Methods
  static Future<Map<String, dynamic>> getIdentityVerification() async {
    return await _makeRequest('GET', '/users/identity-verification');
  }

  static Future<Map<String, dynamic>> submitIdentityVerification({
    required String aadhaarNumber,
    required String panNumber,
    required String aadhaarImagePath,
    required String panImagePath,
  }) async {
    try {
      final token = await _getAccessToken();
      if (token == null) {
        return {'success': false, 'error': 'Not authenticated'};
      }

      final uri = Uri.parse('$baseUrl/users/identity-verification');
      final request = http.MultipartRequest('POST', uri);
      
      request.headers['Authorization'] = 'Bearer $token';
      request.fields['aadhaarNumber'] = aadhaarNumber;
      request.fields['panNumber'] = panNumber;
      
      // Add Aadhaar image
      request.files.add(await http.MultipartFile.fromPath(
        'aadhaarImage',
        aadhaarImagePath,
        contentType: MediaType('image', 'jpeg'),
      ));
      
      // Add PAN image
      request.files.add(await http.MultipartFile.fromPath(
        'panImage',
        panImagePath,
        contentType: MediaType('image', 'jpeg'),
      ));

      final streamedResponse = await request.send();
      final response = await http.Response.fromStream(streamedResponse);
      
      return json.decode(response.body) as Map<String, dynamic>;
    } catch (e) {
      return {'success': false, 'error': e.toString()};
    }
  }

  // Data Download Methods
  static Future<Map<String, dynamic>> requestDataDownload() async {
    return await _makeRequest('POST', '/users/request-data-download');
  }

  static Future<Map<String, dynamic>> getDataDownloadStatus() async {
    return await _makeRequest('GET', '/users/data-download-status');
  }

  // Account Deletion Methods
  static Future<Map<String, dynamic>> submitDeleteAccountRequest({
    required String reason,
    required String feedback,
    required String password,
  }) async {
    return await _makeRequest(
      'POST',
      '/users/delete-account-request',
      body: {
        'reason': reason,
        'feedback': feedback,
        'password': password,
      },
    );
  }

  // Search History Methods
  static Future<Map<String, dynamic>> saveSearchHistory(
    List<Map<String, dynamic>> history,
  ) async {
    return await _makeRequest(
      'POST',
      '/users/search-history',
      body: {'history': history},
    );
  }

  static Future<Map<String, dynamic>> getSearchHistory() async {
    return await _makeRequest('GET', '/users/search-history');
  }

  static Future<Map<String, dynamic>> clearSearchHistory() async {
    return await _makeRequest('DELETE', '/users/search-history');
  }

  // AI Chat Methods
  static Future<Map<String, dynamic>> sendAIMessage({
    required String message,
    List<Map<String, String>>? conversationHistory,
    String? systemPrompt,
    Map<String, dynamic>? options,
  }) async {
    return await _makeRequest(
      'POST',
      '/ai/chat',
      body: {
        'message': message,
        if (conversationHistory != null) 'conversationHistory': conversationHistory,
        if (systemPrompt != null) 'systemPrompt': systemPrompt,
        if (options != null) 'options': options,
      },
    );
  }

  static Future<Map<String, dynamic>> getPropertySuggestions({
    required String budget,
    required String propertyType,
    required String purpose,
    String? preferredArea,
    String? bedrooms,
  }) async {
    return await _makeRequest(
      'POST',
      '/ai/property-suggestions',
      body: {
        'budget': budget,
        'propertyType': propertyType,
        'purpose': purpose,
        if (preferredArea != null) 'preferredArea': preferredArea,
        if (bedrooms != null) 'bedrooms': bedrooms,
      },
    );
  }

  static Future<Map<String, dynamic>> compareAreas({
    required String area1,
    required String area2,
  }) async {
    return await _makeRequest(
      'POST',
      '/ai/compare-areas',
      body: {
        'area1': area1,
        'area2': area2,
      },
    );
  }

  static Future<Map<String, dynamic>> getPriceGuidance({
    required String propertyType,
    required String area,
    required String size,
  }) async {
    return await _makeRequest(
      'POST',
      '/ai/price-guidance',
      body: {
        'propertyType': propertyType,
        'area': area,
        'size': size,
      },
    );
  }

  static Future<Map<String, dynamic>> getAIRateLimitStatus() async {
    return await _makeRequest('GET', '/ai/rate-limit-status');
  }

  static Future<Map<String, dynamic>> getAIModelsInfo() async {
    return await _makeRequest('GET', '/ai/models-info', requireAuth: false);
  }
}

